# Hypothesis Testing Tutorial

In [1]:
import numpy as np
import pandas as pd
from scipy.stats import ttest_1samp, ttest_ind
from statsmodels.stats.weightstats import ztest
from statsmodels.stats.proportion import proportions_ztest

## Problem 1: (20 points)

We have the potato yield from 12 different farms. We know that the standard potato yield for the given variety is µ=20.

x = [21.5, 24.5, 18.5, 17.2, 14.5, 23.2, 22.1, 20.5, 19.4, 18.1, 24.1, 18.5]

Test if the potato yield from these farms is significantly better than the standard yield.


**Question 1.1.** Which type of hypothesis test you will use?

One sample, right sided, t-test

**Question 1.2.** Define the Null and Alternate Hypothesis

H0: $\mu_{potato} = 20$

H1: $\mu_{potato} > 20$

**Question 1.3.** Python implementation.

In [2]:
x = np.array([21.5, 24.5, 18.5, 17.2, 14.5, 23.2, 22.1, 
     20.5, 19.4, 18.1, 24.1, 18.5])
x_mean = np.mean(x)
print(x_mean)
tset, pval = ttest_1samp(x, 20)
print("p-values",pval)
if pval < 0.05:    # alpha value is 0.05 or 5%
    print("We are rejecting null hypothesis")
else:
    print("We are accepting null hypothesis")

20.175
p-values 0.8446291893053613
We are accepting null hypothesis


**Question 1.4.** What is the conclusion? If the significant level is 0.05, reject the null hypothesis or not?

As our p-value is significantly greater than 0.05, we fail to reject the null hypothesis that potato yield from these farms are not significantly better than the standard yield

## Problem 2: (20 points)

We collect the height of a study group which include 7 men and 7 women. We want to know whether the average height of men and women in this group is different. 

First, we import the dataset.

In [3]:
data = 'https://gist.githubusercontent.com/baskaufs/1a7a995c1b25d6e88b45/raw/4bb17ccc5c1e62c27627833a4f25380f27d30b35/t-test.csv'
df = pd.read_csv(data)

df.head()

grouping  height
0      men   181.5
1      men   187.3
2      men   175.3
3      men   178.3
4      men   169.0

**Question 2.1.**  Which type of hypothesis test you will use? Why? Then define the null and alternative hypothesis.

Two sample independent t-test

H0: $\mu_{men} - \mu_{women} = 0$

H1: $\mu_{men} -  \mu_{women} \neq 0$

**Question 2.2.** Before calculating some descriptive statistics, we will subset the data. Create an array `men_height` which extract the height of men samples from the dataframe. Create an array `women_height` which extract the height of women samples from the dataframe.

In [4]:
men_true = df.grouping=='men'
women_true = np.logical_not(men_true)

men_height = df.height.to_numpy()[men_true]
women_height = df.height.to_numpy()[women_true]

# finding the mean
men_mean = np.mean(men_height)
women_mean = np.mean(women_height)

print('Men mean: '+str(men_mean))
print('Women mean: '+str(women_mean))

# finding the standard deviation
men_std = np.std(men_height)
women_std = np.std(women_height)

print('Men std: '+str(men_std))
print('Women std: '+str(women_std))

Men mean: 179.8714285714286
Women mean: 171.05714285714285
Men std: 5.755671559604561
Women std: 5.274969774553117


**Question 2.3.** What is the mean and standard deviation of each group?

$\mu_{\text{men}} = 179.87$, $\mu_{\text{women}} = 171.06$

$\sigma_{\text{men}} = 5.756$, $\sigma_{\text{women}} = 5.275$

**Question 2.4.** Hypothesis test python implementation.

In [5]:
tset,pval = ttest_ind(a=men_height, b=women_height, equal_var=True)
print("p-values",pval)
if pval < 0.05:    # alpha value is 0.05 or 5%
    print("we are rejecting null hypothesis")
else:
    print("we are accepting null hypothesis")

p-values 0.017105701528166175
we are rejecting null hypothesis


**Question 2.5.** What is the conclusion? If the significant level is 0.05, reject the null hypothesis or not?

As our p-value is less than 0.05 we reject the null hypothesis that  the average height of men and women in this group is not different.

## Problem 3: (15 points)

In the population, the average IQ is 100 with a **standard deviation of 15**. A team of students want to test whether their average IQ is larger than 100. They collect the data and save all of them in a csv file.

First, we import the data.

In [6]:
IQ_df = pd.read_csv("gpa_iq.csv")
IQ_df.head()

obs    gpa   iq  gender  concept
0    1  7.940  111       2       67
1    2  8.292  107       2       43
2    3  4.643  100       2       52
3    4  7.470  107       2       66
4    5  8.882  114       1       58

**Question 3.1.**  Which type of hypothesis test you will use? Why? Then define the null and alternative hypothesis.

One sample, right sided, z-test

$\mu_c = 100$


H0: $\mu_{IQ} = \mu_c$

H1: $\mu_{IQ} > \mu_c$

**Question 3.2.** Create an array named `IQ` which extract the IQ values from the dataframe. What is the mean and variance of `IQ` of this team of students?

In [7]:
IQ = IQ_df.iq.to_numpy()

IQ_mean = np.mean(IQ)
IQ_std = np.std(IQ)

print('IQ mean is: '+str(IQ_mean))
print('IQ std is: '+str(IQ_std))

IQ mean is: 108.92307692307692
IQ std is: 13.086271168543274


**Question 3.3.** Hypothesis test python implementation.

In [8]:
tset,pval = ztest(IQ, value = 100, alternative = "larger")
print("p-values",pval)
if pval < 0.05:    # alpha value is 0.05 or 5%
  print("we are rejecting null hypothesis")
else:
  print("we are accepting null hypothesis")

p-values 1.093001671387625e-09
we are rejecting null hypothesis


**Question 3.4.** What is the conclusion? Reject the null hypothesis or not?

Since our p-value is less than 0.05 we are rejecting the null hypothesis that the average IQ of this group of students is larger than 100.

## Problem 4: (15 points)

Let’s suppose that a student wants to figure out if biology professors or English professors know more memes. The student writes a meme quiz and springs it on 14 unsuspecting biology professors and 18 unsuspecting English professors during office hours.

The biology professors get the following scores:
$$3,7,11,0,7,0,4,5,6,2,4,7,2,9$$
and the English professors score:
$$5,5,4,5,4,5,7,2,6,2,2,7,2,6,4,2,5,2$$
We’ll assume that the population variance of the biology professor scores is $σ^2_1=3$
and the population variance of the English professor scores is $σ_2^2=2$.

First, we create the data for this problem.

In [9]:
biology = np.array([3,7,11,0,7,0,4,5,6,2,4,7,2,9])
english = np.array([5,5,4,5,4,5,7,2,6,2,2,7,2,6,4,2,5,2])

**Question 4.1.**  Which type of hypothesis test you will use? Why? Then define the null and alternative hypothesis.

Two sample, two sided, z-test

H0: $\mu_{bio} = \mu_{engl}$

H1: $\mu_{bio} \neq \mu_{engl}$

**Question 4.2.** What is the mean and standard deviation of each group?

In [10]:
bio_mean = np.mean(biology)
engl_mean = np.mean(english)

print('Biology mean: '+str(bio_mean))
print('English mean: '+str(engl_mean))

bio_std = np.std(biology)
engl_std = np.std(english)

print('Biology std: '+str(bio_std))
print('English std: '+str(engl_std))

Biology mean: 4.785714285714286
English mean: 4.166666666666667
Biology std: 3.1436687263807697
English std: 1.740051084818425


**Question 4.3.** Hypothesis test python implementation.

In [11]:
tset,pval = ztest(biology, english, alternative='two-sided')
print("p-values",pval)
if pval < 0.05:    # alpha value is 0.05 or 5%
    print("we are rejecting null hypothesis")
else:
    print("we are accepting null hypothesis")

p-values 0.49324286113932003
we are accepting null hypothesis


**Question 4.4.** What is the conclusion? Take $\alpha=0.05$ as the significant level. Reject the null hypothesis or not?

We are accepting the null hypothesis that Biology Teachers and English Teachers know the same ammount of memes

# Problem 5: (15 points)

Suppose we want to know whether or not the proportion of residents in a certain county who support a certain law is equal to 60%. To test this, we collect the following data on a random sample:

* p0: hypothesized population proportion = 0.60
* x: residents who support law: 64
* n: sample size = 100

**Question 5.1.**  Which type of hypothesis test you will use? Why? Then define the null and alternative hypothesis.

One sample proportion test

H0: $p_s = 0.60$

H1: $p_s \neq 0.60$

\begin{equation}
    z = \frac{p_s - p_0}{\sqrt{\frac{p_s(1 - p_0)}{n}}}
\end{equation}

**Question 5.2.** Hypothesis test python implementation. Calculate p value.

In [12]:
n = 100 # sample size
x = 64 # residents who support the law
p_0 = 0.60 # hypothesized population proportion
p_s = x/n

In [13]:
import statsmodels.api as sm

t_stat, p_value = sm.stats.proportions_ztest(count=x, nobs=n, value= 0.60, alternative='two-sided')
print('Our p-value is: '+str(p_value))

Our p-value is: 0.4046567619272856


**Question 5.3.** What's your conclusion? Take $\alpha=0.05$ as the significant level.

As our p-value is above 0.05, we fail to reject the null hypothesis that the proportion of residents in a certain county who support a certain law is equal to 60%

# Problem 6: (15 points)
Let’s say you’re testing two flu drugs A and B. Drug A works on 41 people out of a sample of 195. Drug B works on 351 people in a sample of 605. Are the two drugs comparable? Use a 5% alpha level.

**Question 6.1.**  Which type of hypothesis test you will use? Why? Then define the null and alternative hypothesis.

Two sample proportion test

H0: $p_{A} - p_{B} = 0$

H1: $ p_{A} - p_{B} \neq 0$

**Question 6.2.** Hypothesis test python implementation. Calculate p value.

In [14]:
from statsmodels.stats.proportion import proportions_ztest

In [15]:
sample_success_c, sample_size_c = (41, 195)
sample_success_t, sample_size_t = (351, 605)


counts = np.array([sample_success_c, sample_success_t])
nobs = np.array([sample_size_c, sample_size_t])

stat, p_value = proportions_ztest(count=counts, nobs= nobs,  alternative='two-sided')

print('Our p-value is: '+str(p_value))

Our p-value is: 2.566230446480293e-19


**Question 6.3.** What's your conclusion? Take $\alpha=0.05$ as the significant level.

As our p-value is below 0.05, we reject our null hypothesis that Drug A and Drug B are comparable.